src/main/java/com/artipie/rpm/misc/StorageLock.java

```diff
@@ -0,0 +1,140 @@
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.rpm.misc;
+
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.blocking.BlockingStorage;
+import java.util.UUID;
+import java.util.function.Predicate;
+
+/**
+ * Locker for the storage.
+ * @since 0.9
+ * @todo #230:30min Use this class in RPM update to prohibit parallel update of the same repository:
+ *  add lock before starting the update and release it on terminate.
+ */
+public final class StorageLock {
+
+    /**
+     * Lock file name.
+     */
+    private static final String FNAME = "lock-%s";
+
+    /**
+     * Lock file name pattern.
+     */
+    private static final String PTRN = "%s/lock-[a-z0-9-]{36}";
+
+    /**
+     * Storage.
+     */
+    private final Storage storage;
+
+    /**
+     * Repository key.
+     */
+    private final Key repo;
+
+    /**
+     * Lock file key.
+     */
+    private final Key file;
+
+    /**
+     * Ctor.
+     * @param storage Storage
+     * @param repo Repo key
+     */
+    public StorageLock(final Storage storage, final Key repo) {
+        this.storage = storage;
+        this.repo = repo;
+        this.file = new Key.From(
+            this.repo, String.format(StorageLock.FNAME, UUID.randomUUID().toString())
+        );
+    }
+
+    /**
+     * Locks storage by adding random file to it.
+     * @throws IllegalStateException On error and if storage is already locked
+     */
+    void lock() {
+        final BlockingStorage bsto = new BlockingStorage(this.storage);
+        try {
+            if (this.noLocks(bsto)) {
+                bsto.save(this.file, new byte[]{});
+                if (bsto.list(this.repo).stream().filter(this.lockKeyPredicate()).count() > 1) {
+                    this.release();
+                    throw this.error();
+                }
+            } else {
+                throw this.error();
+            }
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    /**
+     * Releases lock by removing the lock file.
+     */
+    void release() {
+        try {
+            new BlockingStorage(this.storage).delete(this.file);
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(ex);
```

Comment:
@olenagerasimova if such a situation happens, how do we ensure that the repository is not completely broken and cannot be updated ever without human intervention?